### PR TITLE
[SHLWAPI][SDK] Implement SHGetViewStatePropertyBag Part 3

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Follow-up to #5610. Implementing missing features...
JIRA issue: [CORE-9283](https://jira.reactos.org/browse/CORE-9283)

## Proposed changes

- Add `_EnsureWriteBag`, `_PruneMRUTree`, `_GetMRUSize` and `_GetMRUSlots` helper functions.
- Add code into `_GetMRUSlot`, `_FindNearestInheritBag`, and `Write`.
- Add `CLSID_MruPidlList` and `IID_IMruPidlList` definitions into `<shlguid_undoc.h>`.
- Add `IMruPidlList` interface into `<shlobj_undoc.h>`.

### CODE REVIEW

File: dll/win32/shlwapi/propbag.cpp [hunk 1]
```diff
@@ -1412,11 +1431,50 @@ HKEY CViewStatePropertyBag::_GetHKey(DWORD dwVspbFlags)
     return SHGetShellKey(SHKEY_Key_ShellNoRoam | SHKEY_Root_HKCU, NULL, TRUE);
 }
 
+UINT CViewStatePropertyBag::_GetMRUSize(HKEY hKey)
+{
+    DWORD dwValue, cbValue = sizeof(dwValue);
+
+    if (SHGetValueW(hKey, NULL, L"BagMRU Size", NULL, &dwValue, &cbValue) != ERROR_SUCCESS)
+        return 400;
```
Comment (HBelusca) [inline dll/win32/shlwapi/propbag.cpp:1439]: What would this value correspond to?

File: sdk/include/reactos/shlobj_undoc.h [hunk 2]
```diff
@@ -678,6 +678,34 @@ DECLARE_INTERFACE_(IShellBrowserService, IUnknown)
...
+#undef INTERFACE
+
+#ifdef COBJMACROS
+#define IMruPidlList_QueryInterface(T,a,b) (T)->lpVtbl->QueryInterface(T,a,b)
+#define IMruPidlList_AddRef(T) (T)->lpVtbl->AddRef(T)
+#define IMruPidlList_Release(T) (T)->lpVtbl->Release(T)
+#define IMruPidlList_InitList(T,a,b,c) (T)->lpVtbl->InitList(T,a,b,c)
+#define IMruPidlList_UsePidl(T,a,b) (T)->lpVtbl->UsePidl(T,a,b)
+#define IMruPidlList_QueryPidl(T,a,b,c,d) (T)->lpVtbl->QueryPidl(T,a,b,c,d)
+#define IMruPidlList_PruneKids(T,a) (T)->lpVtbl->PruneKids(T,a)
```
Comment (HBelusca) [inline sdk/include/reactos/shlobj_undoc.h:706]: I would suggest doing something to improve readability, like
```suggestion
#define IMruPidlList_QueryInterface(T,a,b)  (T)->lpVtbl->QueryInterface(T,a,b)
#define IMruPidlList_AddRef(T)              (T)->lpVtbl->AddRef(T)
#define IMruPidlList_Release(T)             (T)->lpVtbl->Release(T)
#define IMruPidlList_InitList(T,a,b,c)      (T)->lpVtbl->InitList(T,a,b,c)
#define IMruPidlList_UsePidl(T,a,b)         (T)->lpVtbl->UsePidl(T,a,b)
#define IMruPidlList_QueryPidl(T,a,b,c,d)   (T)->lpVtbl->QueryPidl(T,a,b,c,d)
#define IMruPidlList_PruneKids(T,a)         (T)->lpVtbl->PruneKids(T,a)
```

File: dll/win32/shlwapi/propbag.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp:1439]: This is the default size of MRU (most recently used) list.
The MRU list has empty rooms at beginning.

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp]: Comment added.

Comment (HBelusca) [reply dll/win32/shlwapi/propbag.cpp]: Isn't it possible to get it, then, via some sizeof() of a particular structure?

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp]: I think nope.

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp]: The MRU list of a shell bag has a non-continuous structure.

image

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp]: A shell bag is basically a data set for system management with a complex structure that is difficult for humans to analyze.

**Pull request merged**